homework/promise.js

```diff
@@ -0,0 +1,118 @@
+'use strict';
+
+/* cSpell:disable */
+{
+  function fetchJSONWithPromise(url) {
+    return new Promise((resolve, reject) => {
+      const xhr = new XMLHttpRequest();
+      xhr.open('GET', url);
+      xhr.responseType = 'json';
+      xhr.onload = () => {
+        if (xhr.status < 400) {
+          resolve(xhr.response);
+        } else {
+          reject(new Error(`Network error: ${xhr.status} - ${xhr.statusText}`));
+        }
+      };
+      xhr.onerror = () => reject(new Error('Network request failed'));
+      xhr.send();
+    });
+  }
+
+  function createAndAppend(name, parent, options = {}) {
+    const elem = document.createElement(name);
+    parent.appendChild(elem);
+    Object.keys(options).forEach(key => {
+      const value = options[key];
+      if (key === 'text') {
+        elem.textContent = value;
+      } else {
+        elem.setAttribute(key, value);
+      }
+    });
+    return elem;
+  }
+
+  function removeChildren(element) {
+    while (element.firstChild) {
+      element.removeChild(element.firstChild);
+    }
+  }
+
+  function renderError(error) {
+    const root = document.getElementById('root');
+    createAndAppend('div', root, { text: error.message, class: 'alert-error', role: 'alert' });
+  }
+
+  function app(repos) {
```

Comment:
Hi Nadin - although the logic here works just fine, I think there is still some room for improvement in `app` function. Firstly, if you remember from JavaScript1 about naming conventions(https://github.com/HackYourFuture/fundamentals/blob/master/fundamentals/naming_conventions.md), the function names should describe the logic inside it with a verb. For example, you already have function names like `renderError`, `removeChildren` which are quite descriptive about the function's internals.

Secondly, the `app` function is doing multiple things in it:
1. Creating layout of the application
2. Creating the select dropdown
3. Having a nested function `displayRepoInfo` which in itself is also doing multiple things like displaying repository details and fetching and displaying contributors information.

I think you can split them in separate functions. It is always good to have multiple functions doing one thing only rather than a single function doing too many things.

If you find yourself in a situation where a variable inside a new function isn't available as it was declared in `app`, try passing that as function arguments to the new function. Drop me a message on Slack if you have any questions. 🙂

Reply:
Thanks, I refactored 👍